index.d.ts

```diff
@@ -68,41 +104,27 @@ declare module "hyperapp" {
   // ---------------------------------------------------------------------------
 
   // A dispatched action handles an event in the context of the current state.
-  type Dispatch<S> = (action: Action<S>, payload?: Payload<any>) => void
+  type Dispatch<S> = (dispatchable: Dispatchable<S>, payload?: any) => void
+
+  // A dispatchable entity, when processed, causes a state transition.
+  type Dispatchable<S, P = any> =
+    | S
+    | StateWithEffects<S>
+    | Action<S, P>
+    | ActionWithPayload<S, P>
 
   // An action transforms existing state and/or wraps another action.
-  type Action<S, P = any> = ActionTransform<S, P> | ActionWithPayload<S, P>
-  type ActionTransform<S, P = any> = (
-    state: State<S>,
-    payload: Payload<P>
-  ) => State<S> | StateWithEffects<S> | Action<S>
-  type ActionWithPayload<S, P> = [ActionTransform<S, P>, Payload<P>]
-
-  // A transform carries out the transition from one state to another.
-  type Transform<S, P = any> = (
-    state: State<S> | StateWithEffects<S>,
-    payload?: Payload<P>
-  ) => State<S> | StateWithEffects<S>
-
-  // Application state is accessible in every view, action, and subscription.
-  type State<S> = S
+  type Action<S, P = any> = (state: S, payload: P) => Dispatchable<S>
+  type ActionWithPayload<S, P = any> = [action: Action<S, P>, payload: P]
 
   // State can be associated with a list of effects to run.
-  type StateWithEffects<S, D = any> = [State<S>, ...Effect<S, D>[]]
-
-  type EffectCreator<S, D = any> = (..._: any[]) => Effect<S, D>
+  type StateWithEffects<S, P = any> = [state: S, ...effects: Effect<S, P>[]]
 
   // An effect is an abstraction over an impure process.
-  type Effect<S, D = any> = [Effecter<S, D>, Payload<D>]
+  type Effect<S, P = any> = [effecter: Effecter<S, P>, payload: P]
 
   // An effecter is where side effects and any additional dispatching may occur.
-  type Effecter<S, D> = (
-    dispatch: Dispatch<S>,
-    payload?: Payload<D>
-  ) => void | Promise<void>
-
-  // A payload is data given to an action, effect, or subscription.
-  type Payload<P> = P
+  type Effecter<S, P> = (dispatch: Dispatch<S>, payload: P) => void | Promise<void>
```

Comment:
Why is this `void | Promise<void>`? Hyperapp doesn't know anything about promises.

Reply:
Because it's reasonable to let effecters use async code.  For example, usage of the browser's `fetch()` API.

Reply:
I see. You may want to declare the function as `async` in order to use `await` in it, causing the function to return a Promise implicitly.

Reply:
Yep, exactly !